src/lib.rs

```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
         // SAFETY: 1 <= i, so not [0], so initialized
         (i > 0).then(|| unsafe { self.items.get_unchecked(i).assume_init_ref() })
     }
+
+    /// Iterator over elements of a collection.
+    ///
+    /// It yields all items in an unspecified order.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use ordsearch::OrderedCollection;
+    /// let expected = vec![1, 2, 3, 4, 5];
+    /// let coll = OrderedCollection::from(expected.clone());
+    /// let mut values: Vec<_> = coll.iter().copied().collect();
+    /// values.sort();
+    /// assert_eq!(values, expected);
+    /// ```
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        Iter { coll: self, idx: 0 }
+    }
+}
+
+impl<T: Clone> OrderedCollection<T> {
+    /// Copies all elemenets into a new [`Vec`] in unspecified order
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use ordsearch::OrderedCollection;
+    /// let x = vec![1, 2, 3, 4, 5];
+    /// let coll = OrderedCollection::from(x.clone());
+    /// let mut values: Vec<_> = coll.into_vec();
+    /// values.sort();
+    /// assert_eq!(values, x);
+    /// ```
+    pub fn into_vec(mut self) -> Vec<T> {
+        assert!(!self.items.is_empty());
+
+        let mut items = mem::take(&mut self.items);
+        items.swap_remove(0);
+        // SAFETY: 0-th element already removed, so all initialized
+        unsafe { mem::transmute(items) }
+    }
+}
+
+impl<T: Clone> IntoIterator for OrderedCollection<T> {
+    type Item = T;
+    type IntoIter = alloc::vec::IntoIter<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.into_vec().into_iter()
+    }
+}
+
+struct Iter<'a, T> {
+    coll: &'a OrderedCollection<T>,
+    idx: usize,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.idx += 1;
+        if self.idx < self.coll.items.len() {
+            let value = &self.coll.items[self.idx];
```

Comment:
This can use `.get_unchecked` as well I think for the same safety reasons as below.

Reply:
Let me do some benchmarking to test if it helps in this particular case and if we may get away without adding an unsafe code.

I have a dubious suspicion that because we already test `idx` for iterator termination criteria, the compiler can eliminate duplicating index checks.

Reply:
I tried following implementation of a `next()`:

```rust
    fn next(&mut self) -> Option<Self::Item> {
        self.idx += 1;
        if self.idx < self.coll.items.len() {
            let value = unsafe { self.coll.items.get_unchecked(self.idx) };
            // SAFETY: i > 0, so only initialized items are accessed
            // SAFETY: i < self.coll.items.len() so no out-of-bounds access
            Some(unsafe { value.assume_init_ref() })
        } else {
            None
        }
    }
```

Criterion doesn't see any significant difference:

```
Iteration/Iteration     time:   [329.94 µs 330.98 µs 332.13 µs]
                        change: [+0.0272% +0.7601% +1.5008%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
```

If you put `self.coll.items.len()` into a blackbox preventing compiler optimizations, there is around 3% improvement when using `get_unchecked()` (comparing to the default variant with blackbox) which is consistent with my theory:

```
Iteration/Iteration     time:   [428.03 µs 428.42 µs 428.83 µs]
                        change: [-4.0191% -3.7239% -3.4410%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
```

I'm advocating for not using unsafe in this scenario.

P.S. let me know if you want me to add benchmarks to the PR.

Reply:
btw, quite probably there is also no need for `get_unchecked()` in a `find_gte()` for the very same reason. But it is out of scope of this PR.

Reply:
Nice, thanks for checking!